#SECTION A

###(Question 1)

In [ ]:
install.packages("readxl")
library(readxl)
# install and load the readxl package for future use and manipulation of excel file.

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



Oepning the excel file, I found sales data is in the first sheet. Therefore, I don't need to specify the sheet name or number when reading excel file in the code below.

In [ ]:
fashion_sales_df = read_excel("Fashion.xlsx") # read the excel file into the dataframe
head(fashion_sales_df, n = 10) # specify the number of rows I want for output and display the dataframe

Order ID    Order Date Shipping Date Aging Ship Mode      Product Category
1  FA-2015-1   2015-05-18 2015-05-25     7    Standard Class Fashion         
2  FA-2015-12  2015-06-19 2015-06-29    10    Standard Class Fashion         
3  FA-2015-23  2015-02-17 2015-02-22     5    Standard Class Fashion         
4  FA-2015-34  2015-02-25 2015-03-03     6    Standard Class Fashion         
5  FA-2015-45  2015-07-09 2015-07-11     2    Standard Class Fashion         
6  FA-2015-56  2015-06-05 2015-06-10     5    Standard Class Fashion         
7  FA-2015-67  2015-10-11 2015-10-18     7    Standard Class Fashion         
8  FA-2015-78  2015-02-14 2015-02-17     3    Standard Class Fashion         
9  FA-2015-89  2015-07-30 2015-08-06     7    Standard Class Fashion         
10 FA-2015-100 2015-12-16 2015-12-21     5    Standard Class Fashion         
   Product    Sales Quantity Discount Profit Shipping Cost Order Priority
1  T - Shirts 248   3        0.03     145.68 14.568        High          
2  T - Shirts 248   5        0.02     143.20 14.320        Low           
3  T - Shirts 248   1        0.04     158.08 15.808        High          
4  T - Shirts 248   3        0.04     138.24 13.824        Low           
5  T - Shirts 248   3        0.01     160.56 16.056        Low           
6  T - Shirts 248   3        0.01     160.56 16.056        High          
7  T - Shirts 248   3        0.03     145.68 14.568        Medium        
8  T - Shirts 248   4        0.01     158.08 15.808        High          
9  T - Shirts 248   2        0.03     153.12 15.312        High          
10 T - Shirts 248   3        0.02     153.12 15.312        High

In [ ]:
summary(fashion_sales_df) # get summery statics for each of all columns(entire dataset)

   Order ID           Order Date                   
 Length:851         Min.   :2015-01-01 00:00:00.0  
 Class :character   1st Qu.:2015-04-09 06:00:00.0  
 Mode  :character   Median :2015-07-03 00:00:00.0  
                    Mean   :2015-07-02 18:21:16.5  
                    3rd Qu.:2015-09-26 18:00:00.0  
                    Max.   :2015-12-31 00:00:00.0  
                    NA's   :5                      
 Shipping Date                        Aging         Ship Mode        
 Min.   :2015-01-04 00:00:00.00   Min.   : 1.000   Length:851        
 1st Qu.:2015-04-13 00:00:00.00   1st Qu.: 3.000   Class :character  
 Median :2015-07-08 00:00:00.00   Median : 5.000   Mode  :character  
 Mean   :2015-07-08 00:33:50.54   Mean   : 5.442                     
 3rd Qu.:2015-10-02 00:00:00.00   3rd Qu.: 8.000                     
 Max.   :2016-01-09 00:00:00.00   Max.   :10.000                     
                                                                     
 Product Category     Pr

###(Question 2)

In [ ]:
str(fashion_sales_df) # get data types of each of all columns and size(row : 851, col : 13)

tibble [851 × 13] (S3: tbl_df/tbl/data.frame)
 $ Order ID        : chr [1:851] "FA-2015-1" "FA-2015-12" "FA-2015-23" "FA-2015-34" ...
 $ Order Date      : POSIXct[1:851], format: "2015-05-18" "2015-06-19" ...
 $ Shipping Date   : POSIXct[1:851], format: "2015-05-25" "2015-06-29" ...
 $ Aging           : num [1:851] 7 10 5 6 2 5 7 3 7 5 ...
 $ Ship Mode       : chr [1:851] "Standard Class" "Standard Class" "Standard Class" "Standard Class" ...
 $ Product Category: chr [1:851] "Fashion" "Fashion" "Fashion" "Fashion" ...
 $ Product         : chr [1:851] "T - Shirts" "T - Shirts" "T - Shirts" "T - Shirts" ...
 $ Sales           : num [1:851] 248 248 248 248 248 248 248 248 248 248 ...
 $ Quantity        : num [1:851] 3 5 1 3 3 3 3 4 2 3 ...
 $ Discount        : num [1:851] 0.03 0.02 0.04 0.04 0.01 0.01 0.03 0.01 0.03 0.02 ...
 $ Profit          : num [1:851] 146 143 158 138 161 ...
 $ Shipping Cost   : num [1:851] 14.6 14.3 15.8 13.8 16.1 ...
 $ Order Priority  : chr [1:851] "High" "Low" "

###(Question 3)

In [ ]:
# will check weather each column has missing values
missing_values <- sapply(fashion_sales_df, function(x) sum(is.na(x)))
# apply the function that calculate total number of missing values in each col to the dataset
#sum(is.na(x)) count the total number of missing values(na) of column


print(missing_values)# Output the number of missing values in each column

        Order ID       Order Date    Shipping Date            Aging 
               0                5                0                0 
       Ship Mode Product Category          Product            Sales 
               0                0                0                0 
        Quantity         Discount           Profit    Shipping Cost 
               0                0               26                0 
  Order Priority 
              12 


###(Question 4)

In [ ]:
print("Before operation")# Before
head(fashion_sales_df, n = 5) # display 5 rows

[1] "Before operation"


Order ID   Order Date Shipping Date Aging Ship Mode      Product Category
1 FA-2015-1  2015-05-18 2015-05-25     7    Standard Class Fashion         
2 FA-2015-12 2015-06-19 2015-06-29    10    Standard Class Fashion         
3 FA-2015-23 2015-02-17 2015-02-22     5    Standard Class Fashion         
4 FA-2015-34 2015-02-25 2015-03-03     6    Standard Class Fashion         
5 FA-2015-45 2015-07-09 2015-07-11     2    Standard Class Fashion         
  Product    Sales Quantity Discount Profit Shipping Cost Order Priority
1 T - Shirts 248   3        0.03     145.68 14.568        High          
2 T - Shirts 248   5        0.02     143.20 14.320        Low           
3 T - Shirts 248   1        0.04     158.08 15.808        High          
4 T - Shirts 248   3        0.04     138.24 13.824        Low           
5 T - Shirts 248   3        0.01     160.56 16.056        Low

In [ ]:
# Replace missing values in the 'Profit' column with the mean excluding the missing values of that column
fashion_sales_df$Profit[is.na(fashion_sales_df$Profit)] <- mean(fashion_sales_df$Profit, na.rm = TRUE)

In [ ]:
print("After operation")#  After
head(fashion_sales_df, n = 5)

[1] "After operation"


Order ID   Order Date Shipping Date Aging Ship Mode      Product Category
1 FA-2015-1  2015-05-18 2015-05-25     7    Standard Class Fashion         
2 FA-2015-12 2015-06-19 2015-06-29    10    Standard Class Fashion         
3 FA-2015-23 2015-02-17 2015-02-22     5    Standard Class Fashion         
4 FA-2015-34 2015-02-25 2015-03-03     6    Standard Class Fashion         
5 FA-2015-45 2015-07-09 2015-07-11     2    Standard Class Fashion         
  Product    Sales Quantity Discount Profit Shipping Cost Order Priority
1 T - Shirts 248   3        0.03     145.68 14.568        High          
2 T - Shirts 248   5        0.02     143.20 14.320        Low           
3 T - Shirts 248   1        0.04     158.08 15.808        High          
4 T - Shirts 248   3        0.04     138.24 13.824        Low           
5 T - Shirts 248   3        0.01     160.56 16.056        Low

###(Question 5)

In [ ]:
# Remove rows where 'Order Priority' is missing by selecting and reassigning rows that the Order Priority column has only non-missing values
fashion_sales_df_clean <- fashion_sales_df[!is.na(fashion_sales_df$`Order Priority`), ]

num_removed_row = nrow(fashion_sales_df) - nrow(fashion_sales_df_clean) # substract the number of rows in original dataset by the number of ones updated
print(paste("The number of row removed:",num_removed_row))

[1] "The number of row removed: 12"


In addition, accoding to the question 3, "Order priority" column has 12 missing values. Thefore, I can know rows that have missing values successfuly deleted if 12 rows has been removed.

###(Question 6)

In [ ]:
# I checked that the data type of Order Date and Shipping Date columns are POSIXct types in question 2, which are not the Date type
fashion_sales_df_clean$`Order Date` = as.Date(fashion_sales_df_clean$`Order Date`) # convert all values(column) in Order Date column to Date type
fashion_sales_df_clean$`Shipping Date` = as.Date(fashion_sales_df_clean$`Shipping Date`)
str(fashion_sales_df_clean) # check whether Order Date and Shipping Date columns successfully turned to Date data types -> confirmed successfully converted

tibble [839 × 13] (S3: tbl_df/tbl/data.frame)
 $ Order ID        : chr [1:839] "FA-2015-1" "FA-2015-12" "FA-2015-23" "FA-2015-34" ...
 $ Order Date      : Date[1:839], format: "2015-05-18" "2015-06-19" ...
 $ Shipping Date   : Date[1:839], format: "2015-05-25" "2015-06-29" ...
 $ Aging           : num [1:839] 7 10 5 6 2 5 7 3 7 5 ...
 $ Ship Mode       : chr [1:839] "Standard Class" "Standard Class" "Standard Class" "Standard Class" ...
 $ Product Category: chr [1:839] "Fashion" "Fashion" "Fashion" "Fashion" ...
 $ Product         : chr [1:839] "T - Shirts" "T - Shirts" "T - Shirts" "T - Shirts" ...
 $ Sales           : num [1:839] 248 248 248 248 248 248 248 248 248 248 ...
 $ Quantity        : num [1:839] 3 5 1 3 3 3 3 4 2 3 ...
 $ Discount        : num [1:839] 0.03 0.02 0.04 0.04 0.01 0.01 0.03 0.01 0.03 0.02 ...
 $ Profit          : num [1:839] 146 143 158 138 161 ...
 $ Shipping Cost   : num [1:839] 14.6 14.3 15.8 13.8 16.1 ...
 $ Order Priority  : chr [1:839] "High" "Low" "High" 

###(Question 7)

In [ ]:
fashion_sales_df_clean$`Shipping Duration` <- as.numeric(fashion_sales_df_clean$`Shipping Date` - fashion_sales_df_clean$`Order Date`)
# Other columns don't have units such as "years old", "$" for cost, so convert shipping duration as numeric values to remove "days" notation.
#Add new column specifying the new column name and assign the expression(calculation) that shows the shipping days taken (= difference b/w shipping date and order date)
print(paste("The number of columns:", ncol(fashion_sales_df_clean))) # check total number of col to confirm new col has been added(13 to 14)
head(fashion_sales_df_clean, n = 10)# confirm new col(Shipping Duration) has been added from visualization

[1] "The number of columns: 14"


Order ID    Order Date Shipping Date Aging Ship Mode      Product Category
1  FA-2015-1   2015-05-18 2015-05-25     7    Standard Class Fashion         
2  FA-2015-12  2015-06-19 2015-06-29    10    Standard Class Fashion         
3  FA-2015-23  2015-02-17 2015-02-22     5    Standard Class Fashion         
4  FA-2015-34  2015-02-25 2015-03-03     6    Standard Class Fashion         
5  FA-2015-45  2015-07-09 2015-07-11     2    Standard Class Fashion         
6  FA-2015-56  2015-06-05 2015-06-10     5    Standard Class Fashion         
7  FA-2015-67  2015-10-11 2015-10-18     7    Standard Class Fashion         
8  FA-2015-78  2015-02-14 2015-02-17     3    Standard Class Fashion         
9  FA-2015-89  2015-07-30 2015-08-06     7    Standard Class Fashion         
10 FA-2015-100 2015-12-16 2015-12-21     5    Standard Class Fashion         
   Product    Sales Quantity Discount Profit Shipping Cost Order Priority
1  T - Shirts 248   3        0.03     145.68 14.568        High          
2  T - Shirts 248   5        0.02     143.20 14.320        Low           
3  T - Shirts 248   1        0.04     158.08 15.808        High          
4  T - Shirts 248   3        0.04     138.24 13.824        Low           
5  T - Shirts 248   3        0.01     160.56 16.056        Low           
6  T - Shirts 248   3        0.01     160.56 16.056        High          
7  T - Shirts 248   3        0.03     145.68 14.568        Medium        
8  T - Shirts 248   4        0.01     158.08 15.808        High          
9  T - Shirts 248   2        0.03     153.12 15.312        High          
10 T - Shirts 248   3        0.02     153.12 15.312        High          
   Shipping Duration
1   7               
2  10               
3   5               
4   6               
5   2               
6   5               
7   7               
8   3               
9   7               
10  5

###(Question 8)

In [ ]:
typeof(fashion_sales_df_clean$`Shipping Duration`) # found that the data type of Shipping Duration (result of calculation, date - date) is double

[1] "double"

Therefore,conditional statement using number such as "greater than 7 " can be applied to Shipping duration" column.

In [ ]:
greater_than_7_days = subset(fashion_sales_df_clean, `Shipping Duration` > 7)
# extract the rows that fulfill the condition, which Shipping duration is greater than 7 days
greater_than_7_days # display all data that is greater than 7 days

Order ID     Order Date Shipping Date Aging Ship Mode      Product Category
1   FA-2015-12   2015-06-19 2015-06-29    10    Standard Class Fashion         
2   FA-2015-210  2015-04-11 2015-04-20     9    Standard Class Fashion         
3   FA-2015-221  2015-10-15 2015-10-24     9    Standard Class Fashion         
4   FA-2015-232  2015-07-24 2015-08-01     8    Standard Class Fashion         
5   FA-2015-265  2015-08-11 2015-08-20     9    Standard Class Fashion         
6   FA-2015-287  2015-08-28 2015-09-06     9    Standard Class Fashion         
7   FA-2015-331  2015-09-13 2015-09-22     9    Standard Class Fashion         
8   FA-2015-408  2015-07-25 2015-08-03     9    Standard Class Fashion         
9   FA-2015-419  2015-05-15 2015-05-24     9    Standard Class Fashion         
10  FA-2015-485  2015-08-09 2015-08-19    10    Standard Class Fashion         
11  FA-2015-551  2015-12-16 2015-12-24     8    Standard Class Fashion         
12  FA-2015-683  2015-08-26 2015-09-04     9    Standard Class Fashion         
13  FA-2015-705  2015-04-05 2015-04-13     8    Standard Class Fashion         
14  FA-2015-738  2015-04-29 2015-05-09    10    Standard Class Fashion         
15  FA-2015-881  2015-07-19 2015-07-27     8    Standard Class Fashion         
16  FA-2015-936  2015-05-31 2015-06-08     8    Standard Class Fashion         
17  FA-2015-958  2015-05-11 2015-05-21    10    Standard Class Fashion         
18  FA-2015-1013 2015-03-09 2015-03-19    10    Standard Class Fashion         
19  FA-2015-1024 2015-01-20 2015-01-30    10    Standard Class Fashion         
20  FA-2015-1057 2015-03-26 2015-04-04     9    Standard Class Fashion         
21  FA-2015-1090 2015-04-28 2015-05-08    10    Standard Class Fashion         
22  FA-2015-1189 2015-04-27 2015-05-06     9    Standard Class Fashion         
23  FA-2015-1211 2015-09-13 2015-09-23    10    Standard Class Fashion         
24  FA-2015-1332 2015-09-19 2015-09-27     8    Standard Class Fashion         
25  FA-2015-1343 2015-06-18 2015-06-28    10    Standard Class Fashion         
26  FA-2015-1354 2015-01-06 2015-01-15     9    Standard Class Fashion         
27  FA-2015-1365 2015-12-23 2016-01-01     9    Standard Class Fashion         
28  FA-2015-1376 2015-12-18 2015-12-26     8    Standard Class Fashion         
29  FA-2015-1387 2015-05-12 2015-05-22    10    Standard Class Fashion         
30  FA-2015-1431 2015-08-06 2015-08-14     8    Standard Class Fashion         
⋮   ⋮            ⋮          ⋮             ⋮     ⋮              ⋮               
219 FA-2015-8339 2015-06-05 2015-06-13     8    Standard Class Fashion         
220 FA-2015-8361 2015-10-27 2015-11-06    10    Standard Class Fashion         
221 FA-2015-8372 2015-02-26 2015-03-07     9    Standard Class Fashion         
222 FA-2015-8416 2015-01-29 2015-02-07     9    Standard Class Fashion         
223 FA-2015-8438 2015-05-20 2015-05-30    10    Standard Class Fashion         
224 FA-2015-8515 2015-05-08 2015-05-16     8    Standard Class Fashion         
225 FA-2015-8548 2015-06-30 2015-07-10    10    Standard Class Fashion         
226 FA-2015-8570 2015-08-28 2015-09-05     8    Standard Class Fashion         
227 FA-2015-8592 2015-12-10 2015-12-20    10    Standard Class Fashion         
228 FA-2015-8625 2015-07-03 2015-07-13    10    Standard Class Fashion         
229 FA-2015-8702 2015-02-21 2015-03-01     8    Standard Class Fashion         
230 FA-2015-8713 2015-10-20 2015-10-30    10    Standard Class Fashion         
231 FA-2015-8724 2015-05-24 2015-06-02     9    Standard Class Fashion         
232 FA-2015-8746 2015-08-13 2015-08-21     8    Standard Class Fashion         
233 FA-2015-8779 2015-01-17 2015-01-25     8    Standard Class Fashion         
234 FA-2015-8801 2015-05-29 2015-06-08    10    Standard Class Fashion         
235 FA-2015-8812 2015-02-06 2015-02-14     8    Standard Class Fashion         
236 FA-2015-8823 2015-01-10 2015-01-18     8    Standard Class Fashion         
237 

###(Question 9)

Data cleaning helps converting raw data into tidy data. In data analysis process, the data cleaning is conducted after collecting data for analysis done (Hui, 2023). Raw data isn't directly ready for analysis because it may contain problems such as having missing values or incorrect data types. Missing values cause bias or calculation errors and incorrect data types cause incorrect interpretation or misoperation of data (What Is Data Cleansing & Why Is It Important? | Alteryx, n.d.). To solve these problems and increase the quality and accuracy of data, data cleaning is necessary, and cleaned data will be helpful for more trustworthy analysis and decision-making in later data analysis process. If data cleaning is conducted incorrectly or insufficiently, it leads to wrong decision making and unreliability of data.

For example, "order date" and "shipping date" columns initially had POSIXct type which can contain date and time information. However, these have only date and don't have time information, and shipping duration will be displayed on "days". Hence, it make more sense to convert into Date types, which leads to easier interpretations and higher quality of data. Also, I checked which column has missing values and handled each of them in appropriate methods. In question 4, imputing by average profit make sense and better than removing rows because profit will be used to calculate total profits of all order in the future. This process makes data higher quality because I didn't delete rows, so I can still can get total profits.

(248 words)


### (Question 10)

I have created new column in question 7. To create new column, I used  symbol, $.

That symbol is normaly used when accessing certain column of dataframe like in Question 6. However, when I specify new column name that haven't used in dataset, I can create new column which has that new name. Then I added the result of calculation to the column, which is the diffrence between shipping date and order date with substraction operator, "-". Finally, I used as.numric(), whcih is the function to convert data or column to numeric type(number). This function remove the "days" infromation from "Shopping duration" and make processing easier in the future(i.g. in conditional statement (stated in the paragraph below)). Moreover, other colums in dataframe doesn't have unit information such as "years old", "$" for cost, so it make scence to remove unit information like other columns for data consistency.

I conducted filtering using subset(), which is the function to apply filtering by condition to dataframe. subset() takes dataframe and its conditional statement to filter as arguments. Therefore, to only extract data that shipping duration is greater than 7 days, I specified the dataframe and its conditional statement, which is column, "Shipping Order" > 7 to filter. Additionally, I converted "shipping order" column to numeric data type and checked that using typeof() function, so the code can compare LHS and RHS on "Shipping Order" > 7  without a error.
(231 words)

#SECTION B

###Snippet 1

In [ ]:
#Define age
age = 20

#Check if age is greater than 18 or not(Check weather defined age will be considered to be Adult or Minor)
if (age > 18) #Error 1:
{
    print("Adult")#Error 2 :
}else #error 3:
{
    print("Minor")
}

[1] "Adult"


###Errors:

1. When parsing the code, the intrepreter hit the "unexpected symbol", "a" which is the chracter type however that is supposed to be a bracket.

2. In the if-else structure, else itself is independently exisiting apart from the if(interpreter read independent else that is connected with if then reading as an error).

3. wrong syntax, : is attached with else


###Why my change resolve the issues:
1. By putting round brackets for the conditional statement, intepreter can find conditional statement correctly.
2. By connecting "else" with if block using curly brackets, {} , interpreter can correctly find there is if-else block.
3. Removing unexpected symbol for else helps interpreter can read or parse code correctly

###Snippet 2

In [ ]:
# Define numbers that holds 1 to 5
numbers = list(1,2,3,4,5) #error 1

#Iterate over 1 to 5 and print each of all numbers in numbers
for (number in numbers){ # error 2
  print(number) # error 3
}

[1] 1
[1] 2
[1] 3
[1] 4
[1] 5


###Errors
1. list is initilsed as wrong form, [ ].
2. for statement is lacking a round bracket, () and interpreter found the unexpected symbol, "n" in number instead of the bracket.
3. Code missing the last round bracket of print function, and interpreter detected unpexpected end of input, whcih is syantex error.

###Why my change resolve the issues:

1. By representing the list in a proper way of R code, interpreter can recognize the list correctly.
2. By putting round brackets for the for statement, intepreter can find for statement correctly.
3. By putting colsing bracket, interpreter can find the end of the code correctly.

###Snippet 3

In [ ]:
#Define variable, "numbers" that have numbers from 1 to 10
numbers = 1:10 #Error 1: logical error

#Create empty list
squares = list() # Error 2

#Iterate over 1 to 10 and if the number is even, calculate square of that number and add the result to the list in each iteration
for (x in numbers){ #Error 4
  if (x %% 2 == 0){ #Error 3
   squares = append(squares, x ** 2)
  }
}
print(squares) # display the list of numbers

[[1]]
[1] 4

[[2]]
[1] 16

[[3]]
[1] 36

[[4]]
[1] 64

[[5]]
[1] 100



###Errors:

1. Code, range(1, 11) in python is meant to be from 1 to 10 inclusive, but range(1,11) in R means min number 1 and max number 11 from 1 to 11(Logical error)

2. Unexpected token, "[" is used for a variable.
(List comprehension, which cannot be used in R, is being used to represent for-if structure in R.)

3. if condition contains wrong syantex, which is % and = (an assignment operator, not a logical operator)

4. for statement is lacking a round bracket, () and interpreter found the unexpected symbol instead of the bracket.


###Why my change resolve the issues:

1. By using : syambol,  that hold vector of sequence of numbers, and the variable "numbers" in the code will hold seqential numbers from 1 to 10.

2. There is no list comprehension in R, so create empty list first and getting value that meets condition(only even numbers) and calculate power of 2 and adding to the list in each iteration.

3. By fixing wrong syyntex of division syambol, from % to %%, division correctly works. Also, the conditional statement should not involve assignment operator, and should check LHS and RHS are equivalent or not. Therefore, code correctly works by changing from assignment operator to logical operator, ==.

4. By putting round brackets for the for statement, intepreter can find for statement correctly.


###Snippet 4

In [ ]:

#Check each of all numbers(name these as outer numbers)(0 to 2) is equivalent to all numbers(name these as inner numbers)(0 to 2).
#However, once the inner number that is the same as the outer number has been found,
#the outer number will be changed to the next sequential number and inner interaction continues until inner number that is the same as the outer number has been found
for (i in 0:2){ # Error 1, Error 2
    for (j in 0:2){ # Error 1, Error 2
        if (i == j){
            break
        }
    # display the outer and inner numbers in the current inner loop.
    print(paste("i = ", i, "j = ", j)) #Error 3
    }
     print("This is after the inner loop") # role of boundary each outer loop
}

[1] "This is after the inner loop"
[1] "i =  1 j =  0"
[1] "This is after the inner loop"
[1] "i =  2 j =  0"
[1] "i =  2 j =  1"
[1] "This is after the inner loop"


###Errors:
1. for statement is lacking a round bracket, () and interpreter found the unexpected symbol instead of the bracket.

2. Code, range(3) in python is meant to be from 0 to 2 inclusive, but range(3) in R hold only 3 (Logical error)

3. After interpreter read f in the print function, interpreter found reaching unexpected string constant, ". (interpreter expected there is a variable, f, but it also found ", which is the queation of string, then recognized as an error)

###Why my change resolve the issues:

1. By putting round brackets for the for statement, intepreter can find for statement correctly.

2. By using : symbol for vector that holds sequential numbers, that hold sequence of numbers, and "0:2" will hold seqential numbers from 0 to 2.

3. By using paste() function, concatinate string and value in R, and code correctly works without an error.


### Snippet 5

In [ ]:
#Error 1

#Define 2*2 sized matrix and 2*3 sized matrix
matrix_a = matrix(c(1,2,3,4), nrow = 2, ncol = 2)
matrix_b = matrix(c(5,6,7,8,9,10), nrow = 2, ncol = 3)

#Perform multiplication of two matrix defined above
result = matrix_a %*% matrix_b # Error 2
print(result)


     [,1] [,2] [,3]
[1,]   23   31   39
[2,]   34   46   58


###Errors:

1. nump onject cannot be found because numpy package is not yet installed and moreover, R doesn't have numpy in the first place.

2. This code is meant to calculate the multipication of the two matrices. However, * symbol actually caliculate multipication of each entry, which is wrong syntax for multipication of the two matrices.


###Why my change resolve the issues:

1. Matrix can be used without importing packages or libraries in R, so just deleting the first line, which is library(nump) helps code to work correctly.

2. Using correct syantax for multipication of the two matrices(changing symbol from * to %*%)




#SECTION C

(Question 1)

In [ ]:
# create vectors that represent columns I will be using to create dataframe.
first_v = as.Date(c("2024-01-01","2024-01-02","2024-01-03","2024-01-04","2024-01-05","2024-01-06","2024-01-07","2024-01-08","2024-01-09", "2024-01-10"))
# For date column, I created vector of strings that represent date and converted it to date data type.
second_v = c("sunny", "sunny","rainy","rainy","cloudy", "sunny","sunny","rainy", "cloudy", "sunny")
third_v = c(FALSE,FALSE,TRUE,TRUE,FALSE,FALSE,TRUE,TRUE,FALSE,FALSE)

Hotel_Booking = data.frame(date = first_v, weather = second_v, school_holiday = third_v) #
# create df putting vectors into the data.frame() function as arguments
print(Hotel_Booking)

         date weather school_holiday
1  2024-01-01   sunny          FALSE
2  2024-01-02   sunny          FALSE
3  2024-01-03   rainy           TRUE
4  2024-01-04   rainy           TRUE
5  2024-01-05  cloudy          FALSE
6  2024-01-06   sunny          FALSE
7  2024-01-07   sunny           TRUE
8  2024-01-08   rainy           TRUE
9  2024-01-09  cloudy          FALSE
10 2024-01-10   sunny          FALSE


(Question 2)

In [ ]:
# prepare variables(define weather and school holiday or not) to predict occupancy rates
#and I will create algorithm that predict occupancy rates based on these variables using control flow structure

weather_cond = "sunny"
school_hol = FALSE

occupancy = -1
# set initial occupancy as -1 (unexpected value). In case output shows -1, I can know there is an error in the algorithm below.

#Predict occupancy algorithm
#Flow: check what is weather first and check holiday or not based on the weather and get occupancy.
# Since the assignment says that "any other conditions -> 50 %"", I put else conditions for both weather and holiday checks.

# check weather first
if (weather_cond == "sunny"){
  # based on the weather and school holiday or not, determine the occupancy
  if (school_hol == TRUE){
    occupancy = 90
  }else if (school_hol == FALSE){
    occupancy = 70
  }else{ # any other condition
    occupancy = 50
  }

}else if (weather_cond == "rainy"){
  if (school_hol == TRUE){
    occupancy = 80
  }else if (school_hol == FALSE){
    occupancy = 60
  }else{
    occupancy = 50
  }

}else if (weather_cond == "cloudy"){
  if (school_hol == TRUE){
    occupancy = 75
  }else if (school_hol == FALSE){
    occupancy = 65
  }else{
    occupancy = 50
  }
}else{ # any other condition
  occupancy = 50
}

print(paste("Occupancy:", occupancy,"%")) # display the result, occupancy

[1] "Occupancy: 70 %"


## Reference list

Hui, E. G. M. (2023, April 7). What is Data Analysis Process? Medium. https://medium.com/@gohminghui88/what-is-data-analysis-process-84864779eb5

‌What Is Data Cleansing & Why Is It Important? | Alteryx. (n.d.). Www.alteryx.com. https://www.alteryx.com/glossary/data-cleansing

